File: geminicrypto.js
require("dotenv").config();
const axios = require("axios");
const { MongoClient } = require("mongodb");

// Load environment variables
const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
const TAVILY_API_KEY = process.env.TAVILY_API_KEY;
const MONGO_URI = process.env.MONGO_URI;

// Gemini API configuration
const GEMINI_URL = `https://generativelanguage.googleapis.com/v1beta/models/gemini-pro:generateContent?key=${GEMINI_API_KEY}`;

// Generate follow-up questions using Gemini REST API
async function generateFollowUpQuestions(userQuery) {
    const prompt = `You are an AI expert in finance and cryptocurrency scams. Your task is to analyze crypto projects and determine if they are likely scams. Given the user query: "${userQuery}", generate a list of follow-up questions to gather more information about the project and its creators. Follow-up questions:`;
    
    const response = await axios.post(GEMINI_URL, {
        contents: [{
            parts: [{ text: prompt }]
        }]
    });

    return response.data.candidates[0].content.parts[0].text.trim();
}

// Search Tavily for answers
async function searchTavilyForAnswers(followUpQuestions) {
    const answers = {};
    const questions = followUpQuestions.split("\n");

    for (const question of questions) {
        if (question.trim()) {
            const response = await axios.post("https://api.tavily.com/search", {
                api_key: TAVILY_API_KEY,
                query: question.trim(),
                max_results: 5,
                search_depth: "advanced"
            });
            answers[question.trim()] = response.data;
        }
    }

    return answers;
}

// Generate final analysis using Gemini REST API
async function generateFinalAnalysis(userQuery, followUpAnswers) {
    const prompt = `You are an AI expert in finance and cryptocurrency scams. Your task is to analyze crypto projects and determine if they are likely scams. Given the user query: "${userQuery}" and the following information: ${JSON.stringify(followUpAnswers)}, provide a detailed analysis of whether the crypto project is likely a scam or not. Include evidence and reasoning.`;

    const response = await axios.post(GEMINI_URL, {
        contents: [{
            parts: [{ text: prompt }]
        }]
    });

    return response.data.candidates[0].content.parts[0].text.trim();
}

// Analyze crypto project
async function analyzeCryptoProject(userQuery) {
    const followUpQuestions = await generateFollowUpQuestions(userQuery);
    const followUpAnswers = await searchTavilyForAnswers(followUpQuestions);
    const finalAnalysis = await generateFinalAnalysis(userQuery, followUpAnswers);

    return {
        userQuery,
        followUpQuestions,
        followUpAnswers,
        finalAnalysis
    };
}

// Store analysis in MongoDB
async function storeAnalysisInDb(result) {
    const client = new MongoClient(MONGO_URI, { useNewUrlParser: true, useUnifiedTopology: true });

    try {
        await client.connect();
        const db = client.db("crypto_analysis");
        const collection = db.collection("user_queries");

        result.timestamp = new Date();
        await collection.insertOne(result);
        console.log("Data inserted:", result);
    } catch (err) {
        console.error("Insertion error:", err);
    } finally {
        await client.close();
    }
}

// Main function
async function main() {
    const userQuery = "Is Trump Coin a scam?"; // Replace with user query from database
    const result = await analyzeCryptoProject(userQuery);
    await storeAnalysisInDb(result);
}

// Run the script
main().catch(console.error);

File: geminicrypto.py
#!pip install google-generativeai flask pymongo requests

import google.generativeai as genai
import requests
from pymongo import MongoClient
from datetime import datetime
from google.colab import userdata

#load gemini
GEMINI_API_KEY = userdata.get("GEMINI_API_KEY")
genai.configure(api_key=GEMINI_API_KEY)
model = genai.GenerativeModel('gemini-2.0-flash')

TAVILY_API_KEY = userdata.get("TAVILY_TOKEN")

# Generate the follow-up questions based on the query with prompt engineering
def generate_follow_up_questions(user_query):
    prompt = f"""
    You are an AI expert in finance and cryptocurrency scams. Your task is to analyze crypto projects and determine if they are likely scams. Given the user query: \"{user_query}\", generate a single follow-up question to gather more information about the project and its creators.
    Follow-up questions:
    """
    # I AM CHANGING THIS TO ONLY ONE QUESTION FOR TESTING PURPOSES ("generate a list of follow-up questions")
    response = model.generate_content(prompt)
    follow_up_questions = response.text.strip()
    return follow_up_questions

  # Use Tavily API to search for answers to the questions (this uses too many tokens)
def search_tavily_for_answers(follow_up_questions):
    answers = {}
    for question in follow_up_questions.split("\n"):  # Split questions by newline
        if question.strip():  # Skip empty lines
            url = "https://api.tavily.com/search"
            payload = {
                "api_key": TAVILY_API_KEY,
                "query": question.strip(),
                "max_results": 1,  # Limit to top 5 results
                "search_depth": "basic"  # Use advanced search for deeper results
            }
            response = requests.post(url, json=payload)
            answers[question.strip()] = response.json()
    return answers
  

def analyze_crypto_project(user_query):
    # Step 1: Generate follow-up questions using Gemini
    follow_up_questions = generate_follow_up_questions(user_query)

    # Step 2: Search for answers to follow-up questions using Tavily
    follow_up_answers = search_tavily_for_answers(follow_up_questions)

    # Step 3: Generate final analysis using Gemini
    final_analysis = generate_final_analysis(user_query, follow_up_answers)

    # Step 4: Return results
    return {
        "user_query": user_query,
        "follow_up_questions": follow_up_questions,
        "follow_up_answers": follow_up_answers,
        "final_analysis": final_analysis
    }


from pymongo import MongoClient
from datetime import datetime
from getpass import getpass
import certifi

MONGO_PASS = getpass("Enter your MongoDB password: ")

MONGO_URI = f"mongodb+srv://laurentcousineau:{MONGO_PASS}@cluster0.cindm.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"

client = MongoClient(MONGO_URI, tlsCAFile=certifi.where())

try:
    client.admin.command("ping")
    print("Pinged your deployment. You successfully connected to MongoDB!")
except Exception as e:
    print("Ping failed:", e)

db = client["crypto_analysis"]
collection = db["user_queries"]

def store_analysis_in_db(result):
    result["timestamp"] = datetime.now()
    try:
        collection.insert_one(result)
        print("Data inserted:", result)
    except Exception as err:
        print("Insertion error:", err)

# Test insertion
# test_data = {"test": "hello world"}
# store_analysis_in_db(test_data)

user_query = "Is Trump Coin a scam?" #THIS NEEDS TO BE THE USER QUERY FROM THE DATABASE
result = analyze_crypto_project(user_query)

store_analysis_in_db(result)
